machine_learning/a_star_search.cpp

```diff
@@ -0,0 +1,659 @@
+    // A helper array to evaluate the next state from current state
+    std::vector<std::pair<int, int>> moves = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
+    /**
+     * @brief Finds an empty space in puzzle (in this case; a zero)
+     * @returns a pair indicating integer distances from top and right
+     * respectively, else returns -1, -1
+     */
+    std::pair<int, int> find_zero() {
+        for (int i = 0; i < N; ++i) {
+            for (int j = 0; j < N; ++j) {
+                if (!board[i][j]) {
+                    return {i, j};
+                }
+            }
+        }
+        return {-1, -1};
+    }
+    /**
+     * @brief check whether the index value is bounded within the puzzle area
+     * @param value index for the current board
+     * @returns `true` if index is within the board, else `false`
+     */
+    inline bool in_range(const int value) const {
+        return value >= 0 && value < N;
+    }
+
+ public:
+    /**
+     * @brief get the value from i units from right and j units from left side
+     * of the board
+     * @param i integer denoting ith row
+     * @param j integer denoting column
+     * @returns non-negative integer denoting the value at ith row and jth
+     * column
+     * @returns -1 if invalid i or j position
+     */
+    int get(size_t i, size_t j) const {
+        if (in_range(i) && in_range(j)) {
+            return board[i][j];
+        }
+        return -1;
+    }
+    /**
+     * @brief Returns the current state of the board
+     */
+    std::array<std::array<int, N>, N> get_state() { return board; }
+
+    /**
+     * @brief returns the size of the EightPuzzle (number of row / column)
+     * @return N, the size of the puzzle.
+     */
+    inline size_t get_size() const { return N; }
+    /**
+     * Default constructor for EightPuzzle
+     */
+    EightPuzzle() {
+        for (size_t i = 0; i < N; ++i) {
+            for (size_t j = 0; j < N; ++j) {
+                board[i][j] = ((i * 3 + j + 1) % (N * N));
+            }
+        }
+    }
+    /**
+     * Parameterized Constructor for EightPuzzle
+     * @param init a 2-dimensional array denoting a puzzle configuration
+     */
+    explicit EightPuzzle(const std::array<std::array<int, N>, N> &init)
+        : board(init) {}
+
+    /**
+     * Copy constructor
+     * @param A a reference of an EightPuzzle
+     */
+    EightPuzzle(const EightPuzzle<N> &A) : board(A.board) {}
+
+    /**
+     * @details move constructor
+     * @param A a reference of an EightPuzzle
+     */
+    EightPuzzle(const EightPuzzle<N> &&A) noexcept
+        : board(std::move(A.board)) {}
+    /**
+     * @details Destructor of EightPuzzle
+     */
+    ~EightPuzzle() = default;
+
+    /**
+     * @details Copy assignment operator
+     * @param A a reference of an EightPuzzle
+     */
+    EightPuzzle &operator=(const EightPuzzle &A) {
+        board = A.board;
+        return *this;
+    }
+
+    /**
+     * @details Move assignment operator
```

Comment:
Check other parts of the code for this, too. 🙂
```suggestion
     * @brief Move assignment operator
```